esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,268 @@
+import fs from 'fs';
+import path from 'path';
+import validatePackageName from 'validate-npm-package-name';
+import {ExportField, ExportMapEntry, PackageManifestWithExports, PackageManifest} from './types';
+import {parsePackageImportSpecifier, resolveDependencyManifest} from './util';
+
+// Rarely, a package will ship a broken "browser" package.json entrypoint.
+// Ignore the "browser" entrypoint in those packages.
+const BROKEN_BROWSER_ENTRYPOINT = ['@sheerun/mutationobserver-shim'];
+
+function hasBrokenBrowserEntrypoint(packageName?) {
+  return !!(packageName && BROKEN_BROWSER_ENTRYPOINT.includes(packageName));
+}
+
+function manifestHasTypes(manifest: PackageManifest): boolean {
+  return !!(manifest.types || manifest.typings);
+}
+
+type ResolveManifestEntryOptions = {
+  packageLookupFields?: string[];
+  packageName?: string;
+};
+
+function resolveManifestEntry(
+  manifest: PackageManifest,
+  entry?: string,
+  {packageLookupFields = [], packageName}: ResolveManifestEntryOptions = {},
+): string | undefined {
+  let foundEntrypoint: string | undefined;
+
+  if (manifest.exports) {
+    foundEntrypoint =
+      typeof manifest.exports === 'string'
+        ? manifest.exports
+        : resolveExportMapEntry(manifest.exports['.']);
+
+    if (typeof foundEntrypoint === 'string') {
+      return foundEntrypoint;
+    }
+  }
+
+  foundEntrypoint = [
+    ...packageLookupFields,
+    'browser:module',
+    'module',
+    'main:esnext',
+    'jsnext:main',
+  ]
+    .map((e) => manifest[e])
+    .find(Boolean);
+
+  if (!foundEntrypoint && !hasBrokenBrowserEntrypoint(packageName)) {
```

Comment:
instead of checking for `if (!foundEntrypoint)` here and in every following conditional below, you could just return foundEntrypoint if it was found in the lookup above. Might simplify this function to just "return fast" when you find the right value, instead of trying to collect the right value and only returning at the end.